### ISSUE TITLE
[OAS3] Schemas with external references fail to resolve properly

### ISSUE THREAD

**Issue by tomblench** [swagger-api/swagger-parser#811]
- swagger-parser version: `io.swagger.parser.v3:swagger-parser-v3:2.0.3-SNAPSHOT` latest from Maven Central.

- Code snippet used to load OpenAPI definition:
`OpenAPI oai = new OpenAPIV3Parser().read("/path/to/index.yaml");`

The attached tarball is a minimal sample which exhibits the issue: 
- `index.yaml` uses external `$ref`s to refer to a response, (`components/response-with-reference.yaml`)
- this in turn refers to a schema (`components/schema-with-reference.yaml`)
- this in turn refers to another schema to define the type of items in an array (`components/referent.yaml`)

On running the code and examining the `toString()` output of the `OpenAPI` object obtained (see attached), the `$ref` for `items` becomes `./components/components/referent.yaml`, which is incorrect. This appears to be due to the invocation of `constructRef` in `ExternalRefProcessor#processRefSchema`.

Attempting to run `openapi-generator-cli` with this definition results in the error
```
java.lang.RuntimeException: Could not find ./components/components/referent.yaml on the classpath
...
[main] WARN  io.swagger.v3.parser.OpenAPIV3Parser - Exception while reading:
java.lang.RuntimeException: Unable to load RELATIVE ref: ./components/components/referent.yaml
```

Finally, we would like to understand which path is used to process relative references, is it the root location (the location containing the `index.yaml`), or is it the location containing the current document?

- The [swagger.io documentation](https://swagger.io/docs/specification/using-ref/) states that "Remote Reference...Uses the whole document located on the same server and in the same location". The phrase "same location" is a little unclear here but my interpretation is that it refers to the "current" file, eg the one containing the `$ref`.
- The [OpenAPI 3 Spec](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#relative-references-in-urls) states that "Relative references used in $ref are processed...using the URL of the current document as the base URI".

If the spec means that the current document should be used as the base URI, then why does the `$ref` in `components/response-with-reference.yaml` have to be `./components/schema-with-reference.yaml`? Surely it should just be `schema-with-reference.yaml` since `components` is already on the "current path"?

You can surely see that there is an inconsistency between the way the reference is made from `response-with-reference.yaml` and `schema-with-reference.yaml`, in that to currently get generation to work, one has to contain `components` and the other not.

[oapi-reference-test.tar.zip](https://github.com/swagger-api/swagger-parser/files/2317762/oapi-reference-test.tar.zip)
[parser-output.txt](https://github.com/swagger-api/swagger-parser/files/2317760/parser-output.txt)

**Comment by tomblench** [comment]
@webron that makes perfect sense to me but that doesn't seem to be happening in practice.

If you look at the attached tar you will see that `response-with-reference.yaml` has the contents:
```
---
description: Response with schema containing reference
content:
  application/json:
    schema:
      $ref: "./components/schema-with-reference.yaml"
```
This correctly resolves `./components/schema-with-reference.yaml` _despite_ the fact that `response-with-reference.yaml` is also in the `components` directory. If your assertion were correct then the `$ref` line should be 
```
      $ref: "./schema-with-reference.yaml"
```
but this fails.

**Comment by webron** [comment]
@tomblench Aha, thanks for the clarification, we'll look into it.

**Comment by gracekarina** [comment]
>$ref: "./schema-with-reference.yaml"

This fails, not because parser can't resolve the response with reference file, but because it's trying to solve the schema inside the response with the parent Directory of the index file.
Parser looks for `./schema-with-reference.yaml` but not inside the components folder but outside in the main folder. Parser solves the schema inside the response in the middle of the resolving the ref response it self, so it takes the parent directory of index, instead of the one provided in the $ref. 
meaning he looks for `./schema-with-reference.yaml` in `oapi-reference-test` folder.

```
+ oapi-reference-test
   index.yaml
   + components
     - response-with-reference.yaml
     - schema-with-reference.yaml
     - referent.yaml

**Comment by tomblench** [comment]
@gracekarina I am still confused:

- Both JSON Schema and OpenAPI state that `$ref`s are processed according to [JSON Reference](https://tools.ietf.org/html/draft-pbryan-zyp-json-ref-03#section-4)
- JSON reference states that "Resolution is performed relative to the referring document"

but

- When I make a reference from a schema in `components` to another schema in `components`, I _do not_ need to give `components` as the leading part of the relative reference
- When I make a reference from anything else (in my case, from a response) in `components` to anything else (in my case a schema) in `components`, I _do_ need to give `components` as the leading part of the relative reference.

Furthermore, if all OpenAPI references (leaving schema→schema references aside for one moment) are resolved relative to the root, then I think the spec is incorrect where it states "Relative references used in $ref are processed as per JSON Reference, using the URL _of the current document_ as the base URI. See also the Reference Object"

**Comment by gracekarina** [comment]
Hi, I did a little testing and came up with this result: Would this be the expected result? 
```
openapi: 3.0.0
info:
  title: Test
  license:
    name: Test
    url: http://test.com/test
  version: 0.0.1
externalDocs:
  description: Test
  url: http://test.com/test
servers:
- url: http://test.com/test
paths:
  /:
    get:
      summary: Test
      description: Test
      operationId: getTest
      responses:
        200:
          description: Response with schema containing reference
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/schema-with-reference'
components:
  schemas:
    referent:
      type: object
    schema-with-reference:
      type: object
      properties:
        reference_array:
          type: array
          items:
            $ref: '#/components/schemas/referent'
  responses:
    response-with-reference:
      description: Response with schema containing reference
      content:
        application/json:
          schema:
            $ref: '#/components/schemas/schema-with-reference'

**Comment by tomblench** [comment]
Yeah that looks about right, assuming that copying `response-with-reference` inline to `responses/200` is the "right" thing to do.

**Comment by gracekarina** [comment]
hi @tomblench  please check the new parser's release. If this issue persists let us know throw a new issue. Thanks!

**Comment by tomblench** [comment]
@gracekarina (sorry for commenting on a closed ticket - something I wouldn't normally do)

Thanks for the quick response and fix, unfortunately we have a related with our schema in #834

**Comment by sivagithub2019** [comment]
Hi, Can somebody help me get this. 
is there a way to validate a json request or response file against a given OpenApi json file. 
I am trying to see if i can use OpenAPIV3Parser to do this ?

**Comment by gracekarina** [comment]
@sivagithub2019 can you please add a new issue, and please add an example so we can discuss, this is a closed issue, so It will be hard to follow in here. thanks!

**Comment by sivagithub2019** [comment]
HiI have create this new issue
need to write a java code for validating a json response against a OpenApi json schema · Issue #1126 · swagger-api/swagger-parser

| 
| 
| 
|  |  |

 |

 |
| 
|  | 
need to write a java code for validating a json response against a OpenA...

need to write a java code for validating a json response against a OpenApi json schema. For example the petstore...
 |

 |

 |





your help is much appreciated in this. thanks. 
    On Friday, June 14, 2019, 06:30:48 PM GMT+1, Grace Karina Gonzalez Diaz <notifications@github.com> wrote:  
 
 
@sivagithub2019 can you please add a new issue, and please add an example so we can discuss, this is a closed issue, so It will be hard to follow in here. thanks!
